src/query/storages/fuse/src/operations/mutation/compact_mutator/full_compact_mutator.rs

```diff
@@ -96,17 +93,32 @@ impl FullCompactMutator {
 #[async_trait::async_trait]
 impl TableMutator for FullCompactMutator {
     async fn target_select(&mut self) -> Result<bool> {
-        let snapshot = self.base_snapshot.clone();
+        let snapshot = self.compact_params.base_snapshot.clone();
         let segment_locations = &snapshot.segments;
-        let mut summarys = Vec::new();
 
+        // Blocks that need to be reorganized into new segments.
+        let mut remain_blocks = Vec::new();
         // Read all segments information in parallel.
         let segments_io = SegmentsIO::create(self.ctx.clone(), self.data_accessor.clone());
-        let segments = segments_io.read_segments(segment_locations).await?;
-        for (idx, segment) in segments.iter().enumerate() {
+        let segments = segments_io
+            .read_segments(segment_locations)
+            .await?
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
+
+        let limit = self.compact_params.limit.unwrap_or(segments.len());
+        if limit < segments.len() {
+            for i in limit..segments.len() {
```

Comment:
We should choose the limit segment from the oldest.
It seems `SegmentInfo` should add timestamp and order by the time desc.
Reference:
https://github.com/datafuselabs/databend/blob/dc5e25186f1479dc7560e1eac7d5dfcd78670424/src/query/storages/fuse/src/io/snapshots.rs#L221-L225

Reply:
I don't think we need to add timestamp.
The current processing logic is that the segments generated by insert are placed in front and the segments generated by compact are placed after.
For example
Table t has 3 segments.
```
s1 s2 s3
```
Now execute `optimize table t compact segment limit 2 `. Compact will select s1 and s2, and assume that compact produces a new segment s4. Then s4 will be appended to the back of the segments: 
```
s3 s4
```
If the insert is executed afterwards and a new segment (s5) is generated. The s5 will be appended to the front of the segments:
```
s5 s3 s4
```
If we then execute `optimize table t compact segment limit 2`, s5 and s3 will be selected. 
But if we select by timestamp, s3 and s4 will be selected.

Reply:
> We should choose the limit segment from the oldest. It seems `SegmentInfo` should add timestamp and order by the time desc. Reference:
> 


I don't think we need to add timestamp.
The current processing logic is that the segments generated by insert are placed in front and the segments generated by compact are placed after.
For example
Table t has 3 segments.
```
s1 s2 s3
```
Now execute `optimize table t compact segment limit 2 `. Compact will select s1 and s2, and assume that compact produces a new segment s4. Then s4 will be appended to the back of the segments: 
```
s3 s4
```
If the insert is executed afterwards and a new segment (s5) is generated. The s5 will be appended to the front of the segments:
```
s5 s3 s4
```
If we then execute `optimize table t compact segment limit 2`, s5 and s3 will be selected. 
But if we select by timestamp, s3 and s4 will be selected.